### PR TITLE
[Bug fix] Llama-3.1-8B eval fixes: duplicate-token decode corruption + Blackhole batched prefill

### PR DESCRIPTION
**Pull request by ifabijanic-tt**
## Summary
Two correctness fixes for Llama-3.1-8B under vLLM on Blackhole (P300x2 / device
`P150x4`, TP4) that were corrupting the eval suite. Prefill-only and decode-only
respectively; no change to other models/devices.

## Fix 1 — Duplicate-token decode corruption under vLLM async scheduling
`models/tt_transformers/tt/generator.py`

Under vLLM async scheduling, at decode **reset / mode-switch** steps the host
supplies a one-step-stale last token (host sampling state lags on-device
sampling), which produced deterministic **batch-wide duplicate-token** output.

Fix: when sampling on device with tracing enabled, on a reset/mode-switch keep
the **authoritative device-sampled token per slot** (permuted by `slot_remap` to
follow condense moves), and only take the host token for slots that were *freshly
prefilled* since the last decode submit (tracked via
`_slots_prefilled_since_decode`).

## Fix 2 — Blackhole batched prefill produced non-deterministic garbage
`model_config.py`, `attention.py`, `mlp.py`, `decoder.py`, `mixtral_moe.py`

Batched prefill (concatenating N same-length users into one forward) emitted
non-deterministic garbage on Blackhole. Root cause: `ttnn.experimental.minimal_matmul`
**races at batched-prefill shapes** — used for QKV prefill (seq > 128) and MLP FF2
(seq > 128); short single-user prefills never hit those paths.

Fix: set a `batched` flag through `get_attn_qkv_program_config` /
`get_mlp_ff2_prg_config` and **gate `minimal_matmul` off for batched prefill on
Blackhole**, falling back to the regular matmul (proven deterministic,
bit-identical to sequential, and L1-OK). Re-enables batched prefill for `P150x4`
(= P300x2); `P300` / `P150x8` kept disabled (untested CCL/AGMM paths). Sequential
prefill is unchanged (keeps `minimal_matmul`, no regression).

## Results
- **Determinism:** 0/64 docs differ across repeated runs; batched output
  bit-identical to sequential.
- **Accuracy (batched ON):** IFEval 78.65 (ratio 0.97 PASS), GPQA 30.13 (1.06 PASS).
- **Perf (`simple_text_demo`, batch-32):** TTFT 24.49 ms → 7.22 ms (~3.4×);
  decode throughput is exactly the same as before (fix is prefill-only).

CI run: https://github.com/tenstorrent/tt-shield/actions/runs/27686952562/job/81898877150

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -1303,18 +1409,14 @@ def _decode_forward_no_trace_text(
...
             user_kv_cache = kv_cache[i] if kv_cache is not None else None
-            decode_out = self.model[i].ttnn_decode_forward(
+            tt_logits_i, tt_log_probs_i = self.model[i].ttnn_decode_forward(
                 tt_tokens[i],
                 tt_current_pos[i],
                 rot_mat_idxs=tt_rot_mat_idxs[i],
                 page_table=tt_page_table[i],
                 kv_cache=user_kv_cache,
-                on_device_logits=on_device_sampling,
+                sampling_on_device=sampling_on_device,
```
Comment (vmaksimovicTT) [inline models/tt_transformers/tt/generator.py]: I think this line might be the issue. Did you update this function in model.py to actually accept this argument? (Or if this function is not supposed to accept this arg, then maybe remove it from here?)

Comment (ifabijanic-tt) [reply models/tt_transformers/tt/generator.py]: You are correct. We also have a conflict with Tomasz latest changes on generator.py and model.py which could impair the correctness of this patch. I have contacted him and we will see if this is a larger issue or just renaming patch.

**Comment by ifabijanic-tt** [comment]
Not needed any more, reduced fully functional fix is here: https://github.com/tenstorrent/tt-metal/pull/47359
